functions/venue.js

```diff
@@ -345,14 +345,15 @@ exports.upsertRoom = functions.https.onCall(async (data, context) => {
     throw new HttpsError("not-found", `Venue ${venueId} not found`);
   }
   const docData = doc.data();
+  let rooms = docData.rooms;
 
   if (typeof roomIndex !== "number") {
-    docData.rooms = [...docData.rooms, room];
+    rooms = [...rooms, room];
   } else {
-    docData.rooms[roomIndex] = room;
+    rooms[roomIndex] = room;
   }
 
-  admin.firestore().collection("venues").doc(venueId).update(docData);
+  admin.firestore().collection("venues").doc(venueId).update({ rooms });
```

Comment:
The solution described in https://github.com/sparkletown/sparkle/pull/1300#pullrequestreview-628399237 is a better way to solve this:

> Since we only actually care about updating the `rooms` key, we should be doing something like this:
> 
> ```diff
> -  if (typeof roomIndex !== "number") {
> -    docData.rooms = [...docData.rooms, room];
> -  } else {
> -    docData.rooms[roomIndex] = room;
> -  }
> -
> - admin.firestore().collection("venues").doc(venueId).update(docData);
> 
> + const newRoomsData = typeof roomIndex !== "number" ? [...docData.rooms, room] : docData.rooms.splice(roomIndex, 1, room);
> + admin.firestore().collection("venues").doc(venueId).update({ rooms: newRoomsData });
> ```
> 
> This should tell it to only update the `room` array, rather than overwriting the entire venue config each time. If we wanted to be super safe/sure about this, we would also still wrap the get/update within a transaction to ensure there can never be a race condition here.

Reply:
It would also be good to wrap this whole section of code (from the initial DB read, to this update) in a transaction:

- https://firebase.google.com/docs/firestore/manage-data/transactions

Reply:
`const newRoomsData = typeof roomIndex !== "number" ? [...docData.rooms, room] : docData.rooms.splice(roomIndex, 1, room);`

Here the method `splice()` returns not the modified array in `newRoomsData` but the modified element only. 
We should keep `docData.rooms` array in variable `rooms` and then use `splice()` and after that `rooms`  will have a modified array.

Reply:
> Here the method splice() returns not the modified array in newRoomsData but the modified element only.

Yup, you're correct about that 🤦🏻 

I think I read this as 'returns the array with the replaced elements' or something. Brainfail.

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice#return_value
  - > An array containing the deleted elements.
  
> We should keep `docData.rooms` array in variable rooms and then use `splice()` and after that rooms will have a modified array.

Sounds like a good plan 👌🏻